nixos/system: Support pre-activated images

roberth opened the pull request:
###### Description of changes

Mostly refactoring, except for the last commit which introduces the feature.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
nixos: Extract module for activation script inclusion into toplevel

Allows omission of this functionality through disabledModules, e.g.
for image building.

roberth committed:
nixos: Move installBootLoader to activation script modules

roberth commented:
@ofborg test simple installer.simple switchTest
@ofborg test systemd-boot
@ofborg test nixos-rebuild-specialisations

RaitoBezarius commented on nixos/modules/system/activation/activatable-system.nix:
```diff
@@ -0,0 +1,48 @@
+      mkdir $out/bin
+      substituteAll ${./switch-to-configuration.pl} $out/bin/switch-to-configuration
+      chmod +x $out/bin/switch-to-configuration
+      ${optionalString (pkgs.stdenv.hostPlatform == pkgs.stdenv.buildPlatform) ''
```
Don't you only need to have `perl` available on the `hostPlatform` ?

RaitoBezarius replied:
(or on host and build?)

roberth replied:
`@perl@` is substituted in `switch-to-configuration.pl`.
`substituteAll` obscures that. Would be nice to be explicit about the variables.

RaitoBezarius replied:
Apologies, my comment was about whether you needed to guard on same host platform and build platform to run the Perl check and I was wondering as Perl is like just checking a script if it could run beyond that condition.

roberth replied:
I suppose you could bring a separate perl for that purpose, but I doubt that it'd be worth the effort.

RaitoBezarius replied:
That's fair

roberth committed:
nixos/activatable-system: Make substitutions explicit

This helps with understanding the code.
We might make this not depend on environment variables later.
systemBuilderArgs is a form of global state, which isn't helpful.

roberth commented:
This has uncovered some previously broken tests, but works with my fixes applied:

 - https://github.com/NixOS/nixpkgs/pull/237069
 - https://github.com/NixOS/nixpkgs/pull/237063

Don't know how to fix:

 - https://github.com/NixOS/nixpkgs/issues/237073

roberth committed:
nixos/activatable-system: Make system builder commands env independent

This way it will be easier to reuse in a different context, such as
a separate build of the activation script by itself (TBD).

roberth committed:
nixos/top-level: Remove unused builder variables

These variables were previously used by the activation script
build commands, but are now embedded into those commands for
to improve reusability for an upcoming addition.

roberth committed:
nixos/tests/switch-test.nix: Fix warnings

roberth committed:
nixos/activatable-system: Move legacy variables to top-level

roberth committed:
nixos: Add system.activatable flag for images that are pre-activated

roberth commented:
I wasn't happy about the lack of validation or test, so I did the thing, not just refactor prep anymore. See last commit.

RaitoBezarius commented:
I am following this PR but my bandwidth is a bit low right now, I will try
to give it a proper review later on.

Le mer. 28 juin 2023 à 14:10, Robert Hensing ***@***.***> a
écrit :

> I wasn't happy about the lack of validation or test, so I did the thing,
> not just refactor prep anymore. See last commit.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/237040#issuecomment-1611287614>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACMZRFGC4MIMMMEJMX7AR3XNQNKLANCNFSM6AAAAAAZBWCGJA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

RaitoBezarius reviewed (approved):
Awesome, this looks good to me. (first read)
I will take it for a walk later on actual systems.

roberth commented:
Thanks!
